File: src/welcome_menu.rb
require 'cursor'
require 'util/actions'
require 'common/text_field'
require 'common/title'
require 'yaml'
require File.join(File.dirname(__FILE__), "common", "text_button")

class WelcomeMenu < BakeryWizard::Window
  include Actions
  include Publisher
  include Subscriber
  
  WIDTH, HEIGHT = 420, 600
  
  BUTTON_OFFSET = REL :x => 36, :y => 40
  BG_OFFSET = REL :x => 0, :y => 0
  
  V_PAD = 20
  V_SPAN = 44
  
  def initialize context
    @context = context
    @cursor = Cursor.new
  end
  
  def window= window
    @window = window
    @cursor.window = self
    @background = Gosu::Image.new(self.window, 'media/game_loader_bg.png', false)
    font = Gosu::Font.new(self.window, 'media/hand.ttf', 35)
    File.exists?(Util.last_played_file_name(@context)) && TextButton.new(self, {:x => BUTTON_OFFSET[:x], :y => BUTTON_OFFSET[:y], :z => 1, :dx => 348, :dy => 44, :image => :game_loader}, :resume_game, font).activate
    TextButton.new(self, {:x => BUTTON_OFFSET[:x], :y => BUTTON_OFFSET[:y] + (V_PAD + V_SPAN), :z => 1, :dx => 348, :dy => 44, :image => :game_loader}, :new_game, font).activate
    TextButton.new(self, {:x => BUTTON_OFFSET[:x], :y => BUTTON_OFFSET[:y] + 2*(V_PAD + V_SPAN), :z => 1, :dx => 348, :dy => 44, :image => :game_loader}, :load_or_save_game, font).activate
    TextButton.new(self, {:x => BUTTON_OFFSET[:x], :y => BUTTON_OFFSET[:y] + 3*(V_PAD + V_SPAN), :z => 1, :dx => 348, :dy => 44, :image => :game_loader}, :credits, font).activate
    TextButton.new(self, {:x => BUTTON_OFFSET[:x], :y => BUTTON_OFFSET[:y] + 4*(V_PAD + V_SPAN), :z => 1, :dx => 348, :dy => 44, :image => :game_loader}, :about, font).activate
    TextButton.new(self, {:x => BUTTON_OFFSET[:x], :y => BUTTON_OFFSET[:y] + 5*(V_PAD + V_SPAN), :z => 1, :dx => 348, :dy => 44, :image => :game_loader}, :go_back, font).activate
    TextButton.new(self, {:x => BUTTON_OFFSET[:x], :y => BUTTON_OFFSET[:y] + 6*(V_PAD + V_SPAN), :z => 1, :dx => 348, :dy => 44, :image => :game_loader}, :exit, font).activate
  end
  
  def resume_game
    $wizard.go_to Shop, {:from_file => Util.last_played_file_name(@context)}
  end
  
  def load_or_save_game
    $wizard.go_to SaveLoad
  end
  
  def new_game
    @context.merge!(YAML.load_file(File.join(File.dirname(__FILE__), '..', 'data', 'new-game-data.yml')))
    warehouse_catlog = YAML.load_file(File.join(File.dirname(__FILE__), '..', 'data', 'warehouse-stock.yml'))
    @context[:assets] = []
    @context[:has_asset_ids].each { |asset_id| @context[:assets] << warehouse_catlog[asset_id] }
    $wizard.go_to StoryPlayer
  end
  
  def go_back
    $wizard.go_to PlayerLoader
  end
  
  def credits
    $wizard.go_to Credits
  end
  
  def about
    $wizard.go_to About
  end
  
  def exit
    Kernel.exit 0
  end
  
  def update
    if button_down? Gosu::Button::MsLeft
      publish(Event.new(:left_click, mouse_x, mouse_y))
    elsif button_down? Gosu::Button::MsRight
      publish(Event.new(:right_click, mouse_x, mouse_y))
    end
  end
  
  def draw
    @background.draw(BG_OFFSET[:x], BG_OFFSET[:y], 0)
    @cursor.draw
    for_each_subscriber { |subscriber| subscriber.render}
  end
end

File: src/welcome_menu.py
import bakery_wizard
from util import game_util
from common import text_button

class WelcomeMenu(bakery_wizard.BaseWindow):

    Y_GAP = 80
    BUTTON_IMAGE_PATH = 'game_loader_button.png'

    def load(self, screen):
        self.bg = game_util.load_image('game_loader_bg.png')
        bakery_wizard.BaseWindow.load(self, screen)
        self.initialize_buttons()

    def initialize_buttons(self):
        image = game_util.load_image(WelcomeMenu.BUTTON_IMAGE_PATH)
        self.initialize_button('resume_game_action', 'Resume Game', self.button_xy(image, -3))
        self.initialize_button('new_game_action', 'New Game', self.button_xy(image, -2))
        self.initialize_button('load_save_action', 'Load or Save Game', self.button_xy(image, -1))
        self.initialize_button('credits_action', 'Credits', self.button_xy(image, 0))
        self.initialize_button('about_action', 'About', self.button_xy(image, 1))
        self.initialize_button('go_back_action', 'Go Back', self.button_xy(image, 2))
        self.initialize_button('exit_action', 'Exit', self.button_xy(image, 3))

    def initialize_button(self, action_method, label, point):
        button = text_button.TextButton(self, action_method, self, label = label, x = point[0], y = point[1], image_path = WelcomeMenu.BUTTON_IMAGE_PATH, size = 30)
        self.sprites.add(button)
        self.register(button)

    def button_xy(self, surface, offset):
        xy = self.center_xy(surface)
        return (xy[0], xy[1] + offset*WelcomeMenu.Y_GAP)

    def resume_game_action(self):
        pass

    def new_game_action(self):
        pass

    def load_save_action(self):
        pass

    def credits_action(self):
        pass

    def about_action(self):
        pass

    def go_back_action(self):
        pass

    def exit_action(self):
        pass


